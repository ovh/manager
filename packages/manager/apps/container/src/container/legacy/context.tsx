import React, { createContext, useContext, useState } from 'react';

type Props = {
  isResponsiveSidebarMenuOpen: boolean;
  setIsResponsiveSidebarMenuOpen: (isOpen: boolean) => void;
};

const LegacyContainerContext = createContext<Props>(undefined);

export function LegacyContainerProvider({
  children,
}: {
  children: JSX.Element;
}) {
  const [
    isResponsiveSidebarMenuOpen,
    setIsResponsiveSidebarMenuOpen,
  ] = useState(false);

  return (
    <LegacyContainerContext.Provider
      value={{
        isResponsiveSidebarMenuOpen,
        setIsResponsiveSidebarMenuOpen,
      }}
    >
      {children}
    </LegacyContainerContext.Provider>
  );
}

export function useLegacyContainer() {
  return useContext(LegacyContainerContext);
}
