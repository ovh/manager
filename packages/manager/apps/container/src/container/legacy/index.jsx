import React, {
  useContext,
  useEffect,
  useRef,
  useState,
  Suspense,
} from 'react';

import { plugin, IFrameMessageBus } from '@ovh-ux/shell';
import ApplicationContext from '@/context';
import NavReshuffleBetaAccessModal from '@/container/common/pnr-beta-modal';
import style from './template.module.scss';

import LegacyHeader from './header';

function LegacyContainer() {
  const iframeRef = useRef(null);
  const [iframe, setIframe] = useState(null);
  const [router, setRouter] = useState(null);
  const { shell } = useContext(ApplicationContext);

  useEffect(() => {
    setIframe(iframeRef.current);
    shell.setMessageBus(new IFrameMessageBus(iframeRef.current));
  }, [iframeRef]);

  useEffect(() => {
    const routing = plugin.routing.initRouting(iframeRef.current);
    shell.registerPlugin('routing', routing);
    setRouter(routing.router);
  }, [iframeRef, shell]);

  return (
    <div className={style.managerShell}>
      {router}
      <Suspense fallback="">
        <NavReshuffleBetaAccessModal />
      </Suspense>
      <div className={style.managerShell_header}>
        <LegacyHeader />
      </div>
      <div className={style.managerShell_content}>
        <iframe
          label="app"
          role="document"
          src="about:blank"
          ref={iframeRef}
        ></iframe>
      </div>
      <div className={style.managerShell_footer}></div>
    </div>
  );
}

export default LegacyContainer;
