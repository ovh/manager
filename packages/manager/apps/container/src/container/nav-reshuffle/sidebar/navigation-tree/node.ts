export type NodeRouting = {
  application?: string;
  hash?: string;
};

export type Node = {
  children?: Node[];
  parent?: Node;
  count?: number | boolean;
  id?: string;
  path?: string;
  serviceType?: string | string[];
  translation?: string;
  features?: string[];
  separator?: boolean;
  idAttr?: string;
  hideIfEmpty?: boolean;
  forceVisibility?: boolean;
  url?: string | Record<string, string>;
  isExternal?: boolean;
  routing?: NodeRouting;
  region?: string[];
};
