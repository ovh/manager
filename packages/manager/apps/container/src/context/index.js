export * from './ApplicationProvider.jsx';
export * from './useApplicationContext';

export { default } from './application.context';
