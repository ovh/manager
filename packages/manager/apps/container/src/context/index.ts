export * from './ApplicationProvider';
export * from './useApplicationContext';

export { default } from './application.context';
