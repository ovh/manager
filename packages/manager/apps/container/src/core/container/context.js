import { createContext } from 'react';

const ContainerContext = createContext();

export default ContainerContext;
