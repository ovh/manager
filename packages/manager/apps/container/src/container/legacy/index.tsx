import React, {
  Suspense,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';

import { IFrameMessageBus } from '@ovh-ux/shell';
import { IFrameAppRouter } from '@/core/routing';

import NavReshuffleBetaAccessModal from '@/container/common/pnr-beta-modal';
import ApplicationContext from '@/context';
import { useProgress } from '@/context/progress';
import { LegacyContainerProvider } from './context';
import LegacyHeader from './Header';
import ServerSidebar from './server-sidebar';
import SidebarOverlay from './server-sidebar/SidebarOverlay';
import style from './template.module.scss';
import Progress from '../common/Progress';
import Preloader from '../common/Preloader';
import usePreloader from '../common/Preloader/usePreloader';
import useContainer from '@/core/container';
import useMfaEnrollment from '@/container/mfa-enrollment';
import MfaEnrollment from '@/container/mfa-enrollment/MfaEnrollment';

function LegacyContainer(): JSX.Element {
  const iframeRef = useRef(null);
  const [iframe, setIframe] = useState<HTMLIFrameElement>(null);
  const { betaVersion } = useContainer();

  const { shell } = useContext(ApplicationContext);
  const { isStarted: isProgressAnimating } = useProgress();
  const preloaderVisible = usePreloader(shell, iframe);
  const applications = shell
    .getPlugin('environment')
    .getEnvironment()
    .getApplications();

  const {
    isMfaEnrollmentForced,
    isMfaEnrollmentVisible,
    hideMfaEnrollment,
  } = useMfaEnrollment();

  useEffect(() => {
    setIframe(iframeRef.current);
    shell.setMessageBus(new IFrameMessageBus(iframeRef.current));
  }, [iframeRef]);

  useEffect(() => {
    const tracking = shell.getPlugin('tracking');
    if (betaVersion) {
      tracking.waitForConfig().then(() => {
        tracking.trackMVTest({
          test: '[product-navigation-reshuffle]',
          waveId: 1,
          creation: '[old-nav]',
        });
      });
    }
  }, []);

  return (
    <LegacyContainerProvider>
      <>
        <Progress isAnimating={isProgressAnimating}></Progress>

        <div className={style.managerShell}>
          <Suspense fallback="">
            <NavReshuffleBetaAccessModal />
          </Suspense>
          <div>
            <LegacyHeader />
          </div>
          <div className={style.managerShell_main}>
            <Suspense fallback="">
              <ServerSidebar />
            </Suspense>
            <div className={style.managerShell_content}>
              <SidebarOverlay />
              {isMfaEnrollmentVisible && (
                <Suspense fallback="">
                  <MfaEnrollment
                    forced={isMfaEnrollmentForced}
                    onHide={hideMfaEnrollment}
                  />
                </Suspense>
              )}
              <Preloader visible={preloaderVisible}>
                <>
                  <IFrameAppRouter
                    iframeRef={iframeRef}
                    configuration={applications}
                  />
                  <iframe
                    title="app"
                    role="document"
                    src="about:blank"
                    ref={iframeRef}
                  ></iframe>
                </>
              </Preloader>
            </div>
          </div>
        </div>
      </>
    </LegacyContainerProvider>
  );
}

export default LegacyContainer;
