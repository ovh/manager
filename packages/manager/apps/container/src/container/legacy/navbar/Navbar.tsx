import React, { useCallback, useState } from 'react';

import { Environment } from '@ovh-ux/manager-config';

import Account from './Account';
import Brand from './Brand';
import Hamburger from './HamburgerMenu';
import style from './navbar.module.scss';
import Search from './Search';
import Universes from './Universes';

import useContainer from '@/core/container';
import LanguageMenu from '@/container/common/language';
import modalStyle from '@/container/common/modal.module.scss';
import NavReshuffleSwitchBack from '@/container/common/nav-reshuffle-switch-back';
import Notifications from '@/container/common/notifications-sidebar/NotificationsButton';
import { useShell } from '@/context';
import { useHeader } from '@/context/header';
import { useUniverses } from '@/hooks/useUniverses';

type Props = {
  environment: Environment;
};

function Navbar({ environment }: Props): JSX.Element {
  const shell = useShell();
  const { universe, setUniverse } = useContainer();
  const { getUniverses, getHubUniverse } = useUniverses();
  const [userLocale, setUserLocale] = useState(
    shell.getPlugin('i18n').getLocale(),
  );

  const [searchURL] = useState<string>();
  const [isDropdownOpen, setIsDropdownOpen] = useState<boolean>(false);
  const { setIsNotificationsSidebarVisible } = useHeader();

  const brandClickHandler = useCallback(
    () =>
      shell.getPlugin('tracking').trackClick({
        name: `navbar::entry::logo`,
        type: 'action',
      }),
    [shell],
  );

  const universeClickHandler = useCallback(
    ({ universe }: Partial<Environment>) => {
      shell.getPlugin('tracking').trackClick({
        name: `navbar::entry::${universe}`,
        type: 'action',
      });
      setUniverse(universe);
    },
    [shell],
  );

  return (
    <>
      <div
        className={`${modalStyle.popoverClickAway} ${
          isDropdownOpen ? '' : modalStyle.hidden
        }`}
      ></div>
      <div className={`oui-navbar ${style.navbar}`}>
        <Hamburger universe={universe} universes={getUniverses()} />
        <Brand
          targetURL={getHubUniverse()?.url || '#'}
          onClick={brandClickHandler}
        />
        <Universes
          universe={universe}
          universes={getUniverses()}
          onClick={universeClickHandler}
        />
        <div
          className={`oui-navbar-list oui-navbar-list_aside oui-navbar-list_end ${style.aside}`}
        >
          {searchURL && (
            <div className="oui-navbar-list__item">
              <Search targetURL={searchURL} />
            </div>
          )}
          <div className="oui-navbar-list__item">
            <NavReshuffleSwitchBack
              onChange={(show: boolean) => {
                setIsDropdownOpen(show);
                setIsNotificationsSidebarVisible(false);
              }}
            />
          </div>
          <div className="oui-navbar-list__item">
            <LanguageMenu
              setUserLocale={setUserLocale}
              userLocale={userLocale}
              onChange={(show: boolean) => {
                setIsDropdownOpen(show);
                setIsNotificationsSidebarVisible(false);
              }}
            ></LanguageMenu>
          </div>
          <div className="oui-navbar-list__item">
            <Notifications />
          </div>
          <Account user={environment.getUser()} />
        </div>
      </div>
    </>
  );
}

export default Navbar;
