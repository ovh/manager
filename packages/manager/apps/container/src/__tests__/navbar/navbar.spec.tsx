import React from 'react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { act, waitFor } from '@testing-library/react';
import { Environment, User } from '@ovh-ux/manager-config';
import { renderWithShell } from '../__test-utils__/contextRenders';

import Navbar from '../../container/legacy/navbar/Navbar';
import { LegacyContainerProvider } from '../../container/legacy/context';

const server = setupServer(
  rest.get('/engine/2api/notification', (req, res, ctx) => {
    return res(ctx.json([]));
  }),
  rest.get('/engine/2api/universes', (req, res, ctx) => {
    return res(ctx.json(['hub', 'dedicated', 'web']));
  }),
  rest.get('/engine/2api/configuration', (req, res, ctx) => {
    return res(ctx.json([]));
  }),
  rest.get(
    '/engine/apiv6/me/preferences/manager/NAV_RESHUFFLE_BETA_ACCESS',
    (req, res, ctx) => {
      return res(ctx.json(false));
    },
  ),
  rest.get(
    '/engine/2api/feature/livechat,pnr:betaV1,pnr:betaV2/availability',
    (req, res, ctx) => {
      return res(ctx.json({}));
    },
  ),
);

describe('UI testing of the navbar', () => {
  beforeAll(() => {
    server.listen();
  });

  afterAll(() => server.close());

  it('renders correctly with environment and shell', async () => {
    const user: Partial<User> = {
      firstname: 'Tester',
      name: 'testee',
      supportLevel: {
        level: '1',
      },
      country: 'FR',
    };
    const universe = 'web';

    const environment: Partial<Environment> = {
      getUser: () => user as User,
      getUniverse: () => universe,
      getUserLocale: () => 'fr_FR',
    };

    let render = null;

    await act(async () => {
      render = await renderWithShell(
        <LegacyContainerProvider>
          <Navbar environment={environment as Environment}></Navbar>
        </LegacyContainerProvider>,
        { environment },
      );
    });

    await waitFor(() => {
      expect(render.asFragment()).toMatchSnapshot();
    });
  });
});
