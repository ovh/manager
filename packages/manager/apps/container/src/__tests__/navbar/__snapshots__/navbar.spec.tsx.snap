// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI testing of the navbar renders correctly with environment and shell 1`] = `
<DocumentFragment>
  <div
    class="popoverClickAway hidden"
  />
  <div
    aria-label="navbar_menu_name"
    class="oui-navbar navbar"
    role="navigation"
  >
    <div>
      <button
        aria-expanded="false"
        class="oui-navbar-toggler oui-navbar-toggler_button"
        role="button"
        type="button"
      >
        <span
          class="oui-navbar-toggler__hamburger"
        >
          <span />
          <span />
          <span />
          <span />
        </span>
      </button>
      <div
        aria-expanded="true"
        class="oui-navbar-menu_toggle oui-navbar-menu oui-navbar-menu_fixed false"
        role="menu"
      />
    </div>
    <a
      aria-label="OVHcloud"
      class="oui-navbar__brand"
      href="#"
    >
      <span
        aria-hidden="true"
        class="oui-icon oui-icon-ovh"
      />
    </a>
    <div
      class="oui-navbar-list"
    >
      <div
        class="oui-navbar-list oui-navbar-list_main oui-navbar_desktop-only"
      >
        <span>
          navbar_loading_universes
        </span>
      </div>
    </div>
    <div
      class="oui-navbar-list oui-navbar-list_aside oui-navbar-list_end aside"
    >
      <div
        class="oui-navbar-list__item"
      />
      <div
        class="oui-navbar-list__item"
      >
        <div
          class="oui-navbar-dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="false"
            aria-label="language_change"
            class="oui-navbar-link oui-navbar-link_dropdown "
            title="language_change"
            type="button"
          >
            <span
              class="oui-navbar-link__wrapper"
            >
              <span
                class="oui-navbar-link__text"
              >
                English
              </span>
              <span
                class="oui-icon oui-icon-chevron-down ml-2"
              />
            </span>
          </button>
          <div
            class="oui-navbar-menu__wrapper"
          >
            <div
              class="oui-navbar-menu oui-navbar-menu_fixed oui-navbar-menu_end"
            >
              <div
                class="oui-navbar-list_dropdown p-4"
              >
                <span
                  class="oui-navbar-list__title title"
                >
                  language_change
                </span>
                <ul
                  class="oui-navbar-list"
                >
                  <li
                    class="oui-navbar-list__item"
                  >
                    <button
                      class="oui-navbar-link oui-navbar-link_tertiary p-0"
                      type="button"
                    >
                      Deutsch
                    </button>
                  </li>
                  <li
                    class="oui-navbar-list__item"
                  >
                    <button
                      class="oui-navbar-link oui-navbar-link_tertiary p-0"
                      type="button"
                    >
                      Español
                    </button>
                  </li>
                  <li
                    class="oui-navbar-list__item"
                  >
                    <button
                      class="oui-navbar-link oui-navbar-link_tertiary p-0"
                      type="button"
                    >
                      Français
                    </button>
                  </li>
                  <li
                    class="oui-navbar-list__item"
                  >
                    <button
                      class="oui-navbar-link oui-navbar-link_tertiary p-0"
                      type="button"
                    >
                      Français (Canadien)
                    </button>
                  </li>
                  <li
                    class="oui-navbar-list__item"
                  >
                    <button
                      class="oui-navbar-link oui-navbar-link_tertiary p-0"
                      type="button"
                    >
                      Italiano
                    </button>
                  </li>
                  <li
                    class="oui-navbar-list__item"
                  >
                    <button
                      class="oui-navbar-link oui-navbar-link_tertiary p-0"
                      type="button"
                    >
                      Polski
                    </button>
                  </li>
                  <li
                    class="oui-navbar-list__item"
                  >
                    <button
                      class="oui-navbar-link oui-navbar-link_tertiary p-0"
                      type="button"
                    >
                      Português
                    </button>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="oui-navbar-list__item"
      >
        <button
          aria-label="navbar_notifications"
          class="oui-navbar-link oui-navbar-link_icon oui-navbar-link_tertiary navbarLink"
          role="button"
          title="navbar_notifications"
          type="button"
        >
          <span
            aria-hidden="true"
            class="oui-icon oui-icon-bell"
          />
        </button>
      </div>
      <button
        aria-label="navbar_account"
        class="oui-navbar-link oui-navbar-link_icon oui-navbar-link_tertiary navbarLink"
        role="button"
        type="button"
      >
        <span
          class="oui-navbar-link__wrapper oui-navbar-link__wrapper_border"
        >
          <span
            aria-hidden="true"
            class="oui-icon navbar-oui-icon oui-icon-user"
          />
          <span
            class="oui-navbar-link__text"
          >
            Tester Testee
          </span>
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
